### PR TITLE
RSA AES key wrap mechanism

### PR DESCRIPTION
**Pull request by varder**
Implement RSA AES key wrap mechanism

The mechanism requires AES KWP mechanism CKM_AES_KEY_WRAP_KWP
which is not currently implemented in OP-TEE nor mentioned in
Global Platform specification.

Use the MBedTLS to wrap/unwrap the target key.

The PKCS#11 Specification:
https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.1/cs01/pkcs11-spec-v3.1-cs01.html#_Toc111203390
6.1.23 RSA AES KEY WRAP

https://github.com/OP-TEE/optee_os/issues/5636

### CODE REVIEW

**Review by jforissier — COMMENTED** [review]
For "libmbedtls: Fix cipher_wrap.c for nist_kw":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: lib/libmbedtls/include/mbedtls_config_uta.h [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #define MBEDTLS_CIPHER_C
 #define MBEDTLS_DES_C
 #define MBEDTLS_AES_C
+#define MBEDTLS_NIST_KW_C
```
Comment (jenswi-linaro) [inline lib/libmbedtls/include/mbedtls_config_uta.h:18]: Please make this change in a separate commit.

Comment (varder) [reply lib/libmbedtls/include/mbedtls_config_uta.h:18]: Done

File: ta/pkcs11/src/processing_asymm.c [hunk 2]
```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	res = TEE_AsymmetricEncrypt(proc->tee_op_handle,
+				    tee_attrs, tee_attrs_count,
+				    aes_key_value, aes_key_size,
+				    out_buf, &aes_wraped_size);
+	if (res) {
+		res = tee2pkcs_error(res);
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_setkey(&kw_ctx, MBEDTLS_CIPHER_ID_AES,
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:967]: Here and at line 974: use a specific `int mbedtls_rc` or like for mdebtls_xxx() return value.
At line 970 (and 979), set `rc` (see comment above).
Or straight:
```c
	if (mbedtls_nist_kw_setkey(...)) {
		rc = PKCS11_CKR_ARGUMENTS_BAD;
		goto out;
	}
```

ditto in `unwrap_rsa_aes_key()`

File: ta/pkcs11/src/processing_asymm.c [hunk 3]
```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+		res = PKCS11_CKR_WRAPPED_KEY_INVALID;
+		goto out;
+	}
+
+	res = mbedtls_nist_kw_unwrap(&kw_ctx, MBEDTLS_KW_MODE_KWP,
+				     (uint8_t *)data + rsa_key_size,
+				     wraped_key_size, *out_buf,
+				     &target_key_size, aes_key_size);
+	if (res)
+		res = PKCS11_CKR_WRAPPED_KEY_INVALID;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:1042]: prefer also `goto out` for consistency.

File: ta/pkcs11/src/processing_asymm.c [hunk 4]
```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	TEE_MemFill(aes_key_value, 0, aes_key_size);
+	mbedtls_nist_kw_free(&kw_ctx);
+	return res;
+}
+
+enum pkcs11_rc wrap_data_by_asymm_enc(struct pkcs11_session *session,
+				      void *data, uint32_t data_sz,
+				      void *out_buf, uint32_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:1061]: replace straight with `enum pkcs11_rc rc = ...`

ditto in `unwrap_key_by_asymm()`

**Comment by etienne-lms** [comment]
@varder, would you have a test case to add to optee_test **xtest** regression test?

File: ta/pkcs11/src/processing_asymm.c [hunk 5]
```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+				       void *out_buf, uint32_t *out_sz)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Attribute tee_attrs[2] = { };
+	struct rsa_aes_key_wrap_processing_ctx *ctx = proc->extra_ctx;
+	mbedtls_nist_kw_context kw_ctx;
+	uint8_t aes_key_value[32] = { };
+	uint32_t aes_key_size = ctx->aes_key_bits / 8;
+	uint32_t tee_attrs_count = 0;
+	uint32_t aes_wraped_size = *out_sz;
```
Comment (sa-kib) [inline ta/pkcs11/src/processing_asymm.c:948]: typo?

```suggestion
	uint32_t aes_wrapped_size = *out_sz;
```

**Comment by sa-kib** [comment]
> @varder, would you have a test case to add to optee_test **xtest** regression test?

@etienne-lms this one is on me, I shall provide test case shortly

File: ta/pkcs11/src/processing_asymm.c (hunk 5 above)
Comment (etienne-lms) [reply ta/pkcs11/src/processing_asymm.c:948]: right. Ditto for `wraped_key_size` in `unwrap_rsa_aes_key()`.

Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: ok

File: ta/pkcs11/src/processing_asymm.c (hunk 4 above)
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

File: ta/pkcs11/src/processing_asymm.c (hunk 2 above)
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

File: ta/pkcs11/src/processing_asymm.c [hunk 6]
```diff
@@ -903,3 +933,159 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	mbedtls_nist_kw_init(&kw_ctx);
+	TEE_GenerateRandom(aes_key_value, aes_key_size);
+	res = TEE_AsymmetricEncrypt(proc->tee_op_handle,
+				    tee_attrs, tee_attrs_count,
+				    aes_key_value, aes_key_size,
+				    out_buf, &aes_wraped_size);
+	if (res) {
+		res = tee2pkcs_error(res);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

**Comment by sa-kib** [comment]
> > @varder, would you have a test case to add to optee_test **xtest** regression test?
> 
> @etienne-lms this one is on me, I shall provide test case shortly

here's corresponding test: https://github.com/OP-TEE/optee_test/pull/627

**Comment by jenswi-linaro** [comment]
For the commits:
- "libmbedtls: Fix cipher_wrap.c for nist_kw"
- "libmbedtls: Enable nist_kw "

Please apply:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
For "libmbedtls: Enable nist_kw":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: ta/pkcs11/src/processing.c [hunk 7]
```diff
@@ -990,18 +990,30 @@ enum pkcs11_rc entry_processing_key(struct pkcs11_client *client,
 			goto out;
 
 	} else if (processing_is_tee_asymm(proc_params->id)) {
-		assert(function == PKCS11_FUNCTION_DERIVE);
 
-		rc = init_asymm_operation(session, function, proc_params,
+		rc = init_asymm_operation(session, operation, proc_params,
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:994]: Before this change, `init_asymm_operation()` was called with `PKCS11_FUNCTION_DERIVE` for derivation. Now it is called with `PKCS11_FUNCTION_ENCRYPT` for key derivation. I think this will hurt existing implementation:
https://github.com/OP-TEE/optee_os/blob/3.19.0/ta/pkcs11/src/processing_asymm.c#L223
https://github.com/OP-TEE/optee_os/blob/3.19.0/ta/pkcs11/src/processing_asymm.c#L289
https://github.com/OP-TEE/optee_os/blob/3.19.0/ta/pkcs11/src/processing_asymm.c#L172
Could you cross check that?

**Comment by varder** [comment]
It turned out that to be able to wrap/unwrap an RSA key there is missing
RSA BER encoding/decoding.
Added some fixup for that

https://github.com/OP-TEE/optee_os/pull/5647/commits/ad2fe47a86caaffc6d52b7a14a0e4e746e6319c8

File: ta/pkcs11/src/processing.c (hunk 7 above)
Comment (varder) [reply ta/pkcs11/src/processing.c]: Hi @etienne-lms
Thanks for pointing this out,
init_asymm_operation was moved in switch statement,

https://github.com/OP-TEE/optee_os/pull/5647/commits/00978ca55be13c1da61ec28972c133dc1d67a752

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 8]
```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+	mbedtls_mpi_free(&P);
+	mbedtls_mpi_free(&Q);
+	TEE_Free(buffer);
+	return rc;
+}
+
 enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 			    size_t key_size)
 {
+	uint32_t type = 0;
```
Comment (sa-kib) [inline ta/pkcs11/src/pkcs11_attributes.c:2381]:
```suggestion
	enum pkcs11_key_type type = 0;
```

for consistency

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 9]
```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
 		if (get_attribute_ptr(head, PKCS11_CKA_VALUE, data, sz))
 			return PKCS11_CKR_ARGUMENTS_BAD;
 		break;
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(head);
+		if (type == PKCS11_CKK_RSA)
+			return encode_rsa_private_key_der(head, data, sz);
+		return PKCS11_CKR_GENERAL_ERROR;
```
Comment (sa-kib) [inline ta/pkcs11/src/pkcs11_attributes.c:2511]:
```suggestion
		fallthrough;
```
to get rid of repetitive returns

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 10]
```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+		*sz = 0;
+	}
+
+	return rc;
 }
 
 enum pkcs11_rc get_key_data_to_wrap(struct obj_attrs *head, void **data,
 				    uint32_t *sz)
 {
+	uint32_t type = 0;
```
Comment (sa-kib) [inline ta/pkcs11/src/pkcs11_attributes.c:2501]: same here, `enum pkcs11_key_type`

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 9 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: That will result in compilation warning

ta/pkcs11/src/pkcs11_attributes.c: In function ‘get_key_data_to_wrap’:
ta/pkcs11/src/pkcs11_attributes.c:2513:6: warning: this statement may fall through [-Wimplicit-fallthrough=]

to get rid of repetitive returns
set rc value instead of return
https://github.com/OP-TEE/optee_os/pull/5647/commits/fcdfab34fd6a9f66fe3809e644fd15da69a8dfcf

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 10 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: ok

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 8 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: ok

File: ta/pkcs11/src/processing.c [hunk 11]
```diff
@@ -1268,6 +1268,7 @@ enum pkcs11_rc entry_wrap_key(struct pkcs11_client *client,
 out:
 	release_active_processing(session);
 out_free:
+	TEE_Free(key_data);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing.c:1292]: it is quite tricky to understand that this free is related to  `get_key_data_to_wrap()` behaviour.
I suggest to rename `get_key_data_to_wrap()` to something like `alloc_key_data_to_wrap()` and an explicit inlien comment description, for example in _ta/pkcs11/src/pkcs11_attributes.c_:
```c
/*
 * Get an allocated copy of key data to be wrapped in @head
 * @head: Object attribute where to find key data to be wrapped
 * @data: Output allocated and filled buffer upon success
 * @sz: Key Output data size in byte upon success
 * Return a pkcs11_rv comliant value
 */
enum pkcs11_rc alloc_key_data_to_wrap(struct obj_attrs *head, void **data,
				    uint32_t *sz)
{
    ...
}
```

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 12]
```diff
@@ -2258,19 +2416,127 @@ static enum pkcs11_rc get_secret_key_data(struct obj_attrs *head, void **data,
...
+	}
+
+	return rc;
+}
+
 enum pkcs11_rc get_key_data_to_wrap(struct obj_attrs *head, void **data,
 				    uint32_t *sz)
 {
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	enum pkcs11_key_type type = 0;
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:2524]: `enum pkcs11_key_type type = PKCS11_CKK_UNDEFINED_ID;`

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 13]
```diff
@@ -2226,30 +2229,314 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (key_bits && mbedtls_mpi_bitlen(&n) != key_bits) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&n), 8) / 8;
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:2287]: prefer use `ROUNDUP_DIV()` (several occurences)

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 14]
```diff
@@ -2226,30 +2229,314 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
 			    size_t key_size)
 {
+	enum pkcs11_key_type type = 0;
+
 	switch (get_class(*head)) {
 	case PKCS11_CKO_SECRET_KEY:
 		return set_secret_key_data(head, data, key_size);
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(*head);
+		if (type == PKCS11_CKK_RSA)
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:2388]: can remove `type` and use `if (get_key_type(head) == PKCS11_CKK_RSA)` straight.

Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: fixed

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 13 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/processing_asymm.c [hunk 15]
```diff
@@ -903,3 +933,172 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	const size_t kw_semiblock_len = 8;
+
+	if (ctx->aes_key_bits != 128 &&
+	    ctx->aes_key_bits != 192 &&
+	    ctx->aes_key_bits != 256)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	mbedtls_nist_kw_init(&kw_ctx);
+	TEE_GenerateRandom(aes_key_value, aes_key_size);
+	rc = TEE_AsymmetricEncrypt(proc->tee_op_handle,
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 12 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 16]
```diff
@@ -2258,19 +2416,127 @@ static enum pkcs11_rc get_secret_key_data(struct obj_attrs *head, void **data,
...
+
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
-		if (get_secret_key_data(head, data, sz))
-			return PKCS11_CKR_ARGUMENTS_BAD;
+		rc = get_secret_key_data(head, data, sz);
+		break;
+	case PKCS11_CKO_PRIVATE_KEY:
+		type = get_key_type(head);
+		if (type == PKCS11_CKK_RSA)
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 17]
```diff
@@ -2258,19 +2416,127 @@ static enum pkcs11_rc get_secret_key_data(struct obj_attrs *head, void **data,
...
+
+	*data = buffer;
+	*sz = mbedtls_rc;
+out:
+	mbedtls_pk_free(&pk);
+
+	if (rc) {
+		TEE_Free(buffer);
+		*data = NULL;
+		*sz = 0;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: ok

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 18]
```diff
@@ -2258,19 +2416,127 @@ static enum pkcs11_rc get_secret_key_data(struct obj_attrs *head, void **data,
...
+	void *q = NULL;
+	void *d = NULL;
+	void *e = NULL;
+	uint32_t n_len = 0;
+	uint32_t p_len = 0;
+	uint32_t q_len = 0;
+	uint32_t d_len = 0;
+	uint32_t e_len = 0;
+	uint8_t *buffer = NULL;
+	mbedtls_pk_context pk;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/processing.c (hunk 11 above)
Comment (varder) [reply ta/pkcs11/src/processing.c:1292]: applied,
thanks for providing solution

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 19]
```diff
@@ -2226,15 +2229,170 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+	uint32_t size = 0;
+	uint32_t buffer_size = 0;
+	void *buffer = NULL;
+	mbedtls_pk_context pk;
+	mbedtls_rsa_context *rsa;
+	mbedtls_mpi n;
+	mbedtls_mpi e;
+	mbedtls_mpi d;
+	mbedtls_mpi p;
+	mbedtls_mpi q;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 20]
```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (key_bits && mbedtls_mpi_bitlen(&N) != key_bits) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
+		goto out;
+	}
+
+	size = ROUNDUP(mbedtls_mpi_bitlen(&N), 8) / 8;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 21]
```diff
@@ -2226,25 +2229,286 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+	if (rc && rc != PKCS11_RV_NOT_FOUND)
+		return rc;
+
+	if (remove_empty_attribute(head, PKCS11_CKA_MODULUS) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PUBLIC_EXPONENT) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIVATE_EXPONENT) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIME_1) ||
+	    remove_empty_attribute(head, PKCS11_CKA_PRIME_2)) {
+		return PKCS11_CKR_GENERAL_ERROR;
+	}
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c]: done

File: ta/pkcs11/src/processing_asymm.c [hunk 22]
```diff
@@ -903,3 +933,166 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	return rc;
+}
+
+static enum pkcs11_rc unwrap_rsa_aes_key(struct active_processing *proc,
+					 void *data, uint32_t data_sz,
+					 void **out_buf, uint32_t *out_sz)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	int mbedtls_rc = 0;
+	TEE_OperationInfo info;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

File: ta/pkcs11/src/processing_asymm.c [hunk 23]
```diff
@@ -903,3 +933,166 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	TEE_GetOperationInfo(proc->tee_op_handle, &info);
+	rsa_key_size = info.keySize / 8;
+	wrapped_key_size = data_sz - rsa_key_size;
+
+	*out_buf = TEE_Malloc(aes_key_size, TEE_MALLOC_FILL_ZERO);
+	if (!*out_buf)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	mbedtls_nist_kw_init(&kw_ctx);
+	rc = TEE_AsymmetricDecrypt(proc->tee_op_handle,
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: done

**Comment by varder** [comment]
Hi @ etienne-lms
Can I squash commits?

File: ta/pkcs11/src/pkcs11_attributes.h [hunk 24]
```diff
@@ -188,8 +188,15 @@ enum pkcs11_rc check_attrs_against_modification(struct pkcs11_session *session,
 enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 			    size_t key_size);
 
-enum pkcs11_rc get_key_data_to_wrap(struct obj_attrs *head, void **data,
-				    uint32_t *sz);
+/*
+ * Get an allocated copy of key data to be wrapped in @head
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.h:192]: s/in/from/

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"libmbedtls: Fix cipher_wrap.c for nist_kw"
"libmbedtls: Enable nist_kw"

For the 2 pkcs11 TA commit, few remaining comments but the overall looks good me.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 25]
```diff
@@ -2226,30 +2229,305 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+
+	rsa = mbedtls_pk_rsa(pk);
+	mbedtls_rc = mbedtls_rsa_export(rsa, &n, &p, &q, &d, &e);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	if (key_bits && mbedtls_mpi_bitlen(&n) != key_bits) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c:2282]: Agree PKCS11_CKR_DATA_LEN_RANGE is not correct here,
But we cannot use CKR_UNWRAPPING_KEY_SIZE_RANGE
because it looks like indicates the problem with the key that is used to decrypt a wrapped target key
CKR_UNWRAPPING_KEY_SIZE_RANGE: It indicates that although the requested unwrapping operation could in principle be carried out, this Cryptoki library (or the token) is unable to actually do it because the supplied key’s size is outside the range of key sizes that it can handle.

Looks like CKR_WRAPPED_KEY_LEN_RANGE
is a better choice
CKR_WRAPPED_KEY_LEN_RANGE: This value can only be returned by C_UnwrapKey. It
indicates that the provided wrapped key can be seen to be invalid solely on the basis of its length.
This return value has higher priority than CKR_WRAPPED_KEY_INVALID.

https://docs.oasis-open.org/pkcs11/pkcs11-spec/v3.1/cs01/pkcs11-spec-v3.1-cs01.html#_Toc111203249

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 26]
```diff
@@ -2226,30 +2229,305 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+enum pkcs11_rc alloc_key_data_to_wrap(struct obj_attrs *head, void **data,
+				      uint32_t *sz)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
-		if (get_attribute_ptr(head, PKCS11_CKA_VALUE, data, sz))
-			return PKCS11_CKR_ARGUMENTS_BAD;
+		rc = get_secret_key_data(head, data, sz);
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c:2520]: done

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 25 above)
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c:2282]: done

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 27]
```diff
@@ -2226,30 +2229,305 @@ static enum pkcs11_rc set_secret_key_data(struct obj_attrs **head, void *data,
...
+		goto out;
+	}
+
+	rc = add_attribute(head, PKCS11_CKA_MODULUS, buffer, size);
+	if (rc)
+		goto out;
+
+	size = ROUNDUP_DIV(mbedtls_mpi_bitlen(&e), 8);
+	if (buffer_size < size) {
+		rc = PKCS11_CKR_DATA_LEN_RANGE;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c:2306]: This checks whether user provided wrapped key has inconsistent parameters
if (buffer_size < size) {
I would suggest to use
PKCS11_CKR_WRAPPED_KEY_LEN_RANGE

File: ta/pkcs11/src/processing_asymm.c [hunk 28]
```diff
@@ -903,3 +933,174 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+		rc = PKCS11_CKR_WRAPPED_KEY_INVALID;
+		goto out;
+	}
+
+	*out_sz = target_key_size;
+out:
+	if (rc) {
+		TEE_Free(*out_buf);
+		*out_buf = NULL;
+		*out_sz = 0;
```
[the remark this answers is not in the document]
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c:1064]: ok

File: ta/pkcs11/src/pkcs11_attributes.h (hunk 24 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.h:192]: done

File: ta/pkcs11/src/processing_asymm.c [hunk 29]
```diff
@@ -903,3 +933,168 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_nist_kw_wrap(&kw_ctx, MBEDTLS_KW_MODE_KWP,
+					  data, data_sz,
+					  (uint8_t *)out_buf + aes_wrapped_size,
+					  &target_key_size,
+					  *out_sz - aes_wrapped_size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:990]: here I think we should return `PKCS11_CKR_KEY_SIZE_RANGE` when `mbedtls_rc == MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA`
and return `PKCS11_CKR_KEY_NOT_WRAPPABLE` for other `mbedtls_rc` error codes.

File: ta/pkcs11/src/processing_asymm.c [hunk 30]
```diff
@@ -903,3 +933,168 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+	if (res) {
+		if (res == TEE_ERROR_SHORT_BUFFER)
+			*out_sz = aes_wrapped_size + data_sz + kw_semiblock_len;
+
+		rc = tee2pkcs_error(res);
+		goto out;
+	}
+
+	if (*out_sz < aes_wrapped_size) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:972]: Is this is a case where we should return `ERROR_SHORT_BUFFER` and provide the expected output buffer size in `*out_sz`?

File: ta/pkcs11/src/processing_asymm.c [hunk 31]
```diff
@@ -903,3 +933,168 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+		goto out;
+	}
+
+	mbedtls_rc = mbedtls_nist_kw_setkey(&kw_ctx, MBEDTLS_CIPHER_ID_AES,
+					    aes_key_value, ctx->aes_key_bits,
+					    true);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:980]: This is not expected to failed. I would rathe return `PKCS11_CKR_FUNCTION_FAILED` here, rahter than `ARGUMENTS_BAD`.

File: ta/pkcs11/src/processing_asymm.c [hunk 32]
```diff
@@ -903,3 +933,168 @@ enum pkcs11_rc do_asymm_derivation(struct pkcs11_session *session,
...
+					  data, data_sz,
+					  (uint8_t *)out_buf + aes_wrapped_size,
+					  &target_key_size,
+					  *out_sz - aes_wrapped_size);
+	if (mbedtls_rc) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	*out_sz = target_key_size + aes_wrapped_size;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:994]: here we expect `target_key_size == data_sz + kw_semiblock_len`? otherwise the expect output size returned with `SHORT_BUFFER` is wrong. I suggest to add an assertion here:
`assert(target_key_size <= data_sz + kw_semiblock_len)`

**Review by etienne-lms — COMMENTED** [review]
some more quetions on error code in key wraping sequence.

File: ta/pkcs11/src/processing_asymm.c (hunk 29 above)
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c:996]: ok

File: ta/pkcs11/src/processing_asymm.c (hunk 31 above)
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c]: ok

File: ta/pkcs11/src/processing_asymm.c (hunk 32 above)
Comment (varder) [reply ta/pkcs11/src/processing_asymm.c:1001]: Here we expect
target_key_size <= data_sz + kw_semiblock_len

I suggest to add to check  if we want to make sure there is no overflow have occurred
*out_sz >= expected_size + target_key_size

**Comment by varder** [comment]
Hi @etienne-lms
Can I squash the fixup?

File: ta/pkcs11/src/processing_asymm.c [hunk 33]
```diff
@@ -960,24 +961,29 @@ static enum pkcs11_rc wrap_rsa_aes_key(struct active_processing *proc,
...
 	}
 
 	mbedtls_rc = mbedtls_nist_kw_setkey(&kw_ctx, MBEDTLS_CIPHER_ID_AES,
 					    aes_key_value, ctx->aes_key_bits,
 					    true);
 	if (mbedtls_rc) {
-		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		if (mbedtls_rc == MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA)
+			rc = PKCS11_CKR_KEY_SIZE_RANGE;
+
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_asymm.c:985]: else

**Review by etienne-lms — COMMENTED** [review]
Yes, you can address new review comment and squash the fixup commit.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 34]
```diff
@@ -2258,19 +2412,122 @@ static enum pkcs11_rc alloc_copy_attribute_value(struct obj_attrs *head,
...
 				      uint32_t *sz)
 {
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+
 	switch (get_class(head)) {
 	case PKCS11_CKO_SECRET_KEY:
-		if (alloc_copy_attribute_value(head, data, sz))
-			return PKCS11_CKR_ARGUMENTS_BAD;
+		rc = alloc_copy_attribute_value(head, data, sz);
+		break;
```
Comment (etienne-lms) [inline ta/pkcs11/src/pkcs11_attributes.c:2521]: Actaully this fixes an issue in previous commit "ta: pkcs11: Copy secret key data for wrapping" which returned ARGS_BAD instead of `rc`. Yet, I think it's not probelmatic and ok as is. My A-b stands.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"ta: pkcs11: Copy secret key data for wrapping"
and "ta: pkcs11: Add RSA AES key wrap".
Thanks.

File: ta/pkcs11/src/pkcs11_attributes.c (hunk 34 above)
Comment (varder) [reply ta/pkcs11/src/pkcs11_attributes.c:2521]: Agree, no need to return ARGS_BAD here,
because this check is done already in alloc_copy_attribute_value.
remove repetitive check in "ta: pkcs11: Copy secret key data for wrapping"
